compiler/rustc_typeck/src/check/writeback.rs

```diff
@@ -192,6 +192,34 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         }
     }
 
+    // (ouz-a #100498): Normally `[T] : std::ops::Index<usize>` should be normalized
+    // into [T] but currently `Where` clause stops the normalization process for it,
+    // here we compare types of expr and base in a code without `Where` clause they would be equal
+    // if they are not we don't modify the expr, hence we bypass the ICE
+    fn is_builtin_index(
+        &mut self,
+        typeck_results: &TypeckResults<'tcx>,
+        e: &hir::Expr<'_>,
+        base_ty: Ty<'tcx>,
+        base_span: Span,
+    ) -> bool {
+        if base_ty.builtin_index().is_none() {
+            return false;
+        } else {
+            let resolved_base_ty = self.resolve(base_ty, &base_span);
+            let elem_ty = resolved_base_ty.builtin_index().unwrap();
+
+            let exp_ty = typeck_results.expr_ty(e);
+            let resolved_exp_ty = self.resolve(exp_ty, &e.span);
+
+            if elem_ty != resolved_exp_ty {
```

Comment:
You can return `elem_ty == resolved_exp_ty` directly, and also no need to use the `return` operator here